pkg/tenant/topics.go

```diff
@@ -69,68 +70,101 @@ func (s *Server) ProjectTopicCreate(c *gin.Context) {
 		return
 	}
 
-	// Bind the user request with JSON and return a 400 response
-	// if binding is not successful.
+	// Bind the user request with JSON and return a 400 response if binding is not successful.
 	if err = c.BindJSON(&topic); err != nil {
 		log.Warn().Err(err).Msg("could not bind project topic create request")
 		c.JSON(http.StatusBadRequest, api.ErrorResponse("could not bind request"))
 		return
 	}
 
-	// Verify that a topic ID does not exist and return a 400 response
-	// if the topic ID exists.
+	// Verify that a topic ID does not exist and return a 400 response if the topic ID exists.
 	if topic.ID != "" {
 		c.JSON(http.StatusBadRequest, api.ErrorResponse("topic id cannot be specified on create"))
 		return
 	}
 
-	// Verify that a topic name has been provided and return a 400 response
-	// if the topic name does not exist.
+	// Verify that a topic name has been provided and return a 400 response if the topic name does not exist.
 	if topic.Name == "" {
 		c.JSON(http.StatusBadRequest, api.ErrorResponse("topic name is required"))
 		return
 	}
 
-	t := &db.Topic{
-		OrgID:     orgID,
-		ProjectID: projectID,
-		Name:      topic.Name,
+	// Get project ID from the URL and return a 404 response if it is missing.
+	var projectID ulid.ULID
+	if projectID, err = ulid.Parse(c.Param("projectID")); err != nil {
+		log.Error().Err(err).Str("projectID", projectID.String()).Msg("could not parse project ulid")
+		c.JSON(http.StatusNotFound, api.ErrorResponse("project not found"))
+		return
 	}
 
-	// Add topic to the database and return a 500 response if not successful.
-	if err = db.CreateTopic(c.Request.Context(), t); err != nil {
-		log.Error().Err(err).Msg("could not create project topic in the database")
-		c.JSON(http.StatusInternalServerError, api.ErrorResponse("could not add project topic"))
+	// Retrieve project from the database.
+	var project *db.Project
+	if project, err = db.RetrieveProject(ctx, projectID); err != nil {
+		log.Error().Err(err).Str("projectID", projectID.String()).Msg("could not retrieve project from database")
+		c.JSON(http.StatusNotFound, api.ErrorResponse("project not found"))
 		return
 	}
 
-	c.JSON(http.StatusCreated, t.ToAPI())
+	// Ensure a project cannot be created in a different organization.
+	if claims.OrgID != project.OrgID.String() {
+		log.Error().Err(err).Str("user_orgID", claims.OrgID).Str("project_orgID", project.OrgID.String()).Msg("project org ID does not match user org ID")
+		c.JSON(http.StatusNotFound, api.ErrorResponse("project not found"))
+		return
+	}
+
+	// Get access to the project from Quarterdeck.
+	req := &qd.Project{
+		ProjectID: project.ID,
+	}
+
+	var rep *qd.LoginReply
+	if rep, err = s.quarterdeck.ProjectAccess(ctx, req); err != nil {
+		log.Error().Err(err).Msg("could not get access to project claims")
+		c.JSON(qd.ErrorStatus(err), api.ErrorResponse("could not create topic"))
+		return
+	}
+
+	// Create Ensign context.
+	enCtx := qd.ContextWithToken(ctx, rep.AccessToken)
```

Comment:
We'll probably want to create a parallel story for creating a specific ensign context; `qd` I assume, is the quarterdeck library - which might not work the same as Ensign. 

```suggestion
	// Create Ensign context.
	// TODO: ensure the context has PerRPCCredentials for gRPC authentication
	enCtx := qd.ContextWithToken(ctx, rep.AccessToken)
```